routes/folders.js

```diff
@@ -0,0 +1,110 @@
+const Note = require('../models/note');
+
+
+/* ========== GET/READ ALL FOLDERS ========== */
+folderRouter.get('/', (req, res, next) => {
+   return Folder.find().sort({ name: 'desc' })
+  .then(folders => {
+    res.json(folders);
+  })
+  .catch(err => {
+    console.error(`ERROR: ${err.message}`);
+    console.error(err);
+    return next(err);
+  });
+});
+
+/* ========== GET/READ A SINGLE FOLDERS ========== */
+folderRouter.get('/:id', (req, res, next) => {
+  let id = req.params.id;
+  if(!mongoose.Types.ObjectId.isValid(id)){
+    const err = new Error('id invalid');
+    err.status = 400;
+    return next(err);
+  }
+    return Folder.findById(id)
+    .then(function(result) {
+        result? res.status(200).json(result): res.status(404).json({ "error" : "Id not found"});
+    })
+    .catch(err => 
+    res.status(404).json({ "error" : "Id not *found"}));
+});
+
+/* ========== POST/CREATE A FOLDER ========== */
+folderRouter.post('/', (req, res, next) => {
+  const { name } = req.body;
+  if (!name) {
+    const err = new Error('Missing `name` in request body');
+    err.status = 400;
+    return next(err);
+  }
+  let folder = { name };
+  return Folder.create(folder)
+  .then(function(result) {
+    res.location( `http://${req.headers.host}/api/folders/${result._id}`).status(201).json(result);
+  })
+  .catch(err => {
+    if (err.code === 11000) {
+      err = new Error('The folder name already exists');
+      err.status = 400;
+    }
+    next(err);
+  });
+});
+
+/* ========== PUT/UPDATE A SINGLE FOLDER ========== */
+folderRouter.put('/:id', (req, res, next) => {
+  const id = req.params.id;
+    if(!mongoose.Types.ObjectId.isValid(id)){
+    const err = new Error('id invalid');
+    err.status = 400;
+    return next(err);
+  }
+  const { name } = req.body;
+  if (!name) {
+    const err = new Error('Missing `name` in request body');
+    err.status = 400;
+    return next(err);
+  }     
+  return Folder.findByIdAndUpdate(id, { name, updatedAt : new Date() }, {new: true})
+  .then(function(result) {
+    res.json(result);
+  })
+  .catch(err => {
+    if (err.code === 11000) {
+      err = new Error('The folder name already exists');
+      err.status = 400;
+    }
+    next(err);
+  });
+});
+
+/* ========== DELETE/REMOVE A SINGLE FOLDER ========== */
+//loop thru notes looking for folderId that matches the Folder._id 
+// $unset the id on the notes that match.      line 100 actually worked!.. thought it was going to be harder than that..
+folderRouter.delete('/:id', (req, res, next) => {
+  const id = req.params.id;
+  if(!mongoose.Types.ObjectId.isValid(id)){
+	const err = new Error('id invalid');
+	err.status = 400;
+    return next(err);
+  }
+  return Note.updateMany( {folderId: id },{ $unset : { folderId : "" }})
+    .then(()=>{
+      return Folder.findByIdAndRemove(id)
+        .then(()=>res.status(204).end())
+        .catch(()=> {
+        res.status(404).json({"error": "id not found in database!"});
```

Comment:
prefer error handler
```suggestion
 next(err)
```